scmgr/app.go

```diff
@@ -149,11 +67,125 @@ func main() {
 		log.SetDebugVisible(c.Int("debug"))
 		return nil
 	}
-	cliApp.Run(os.Args)
+	log.ErrFatal(cliApp.Run(os.Args))
 }
 
+// adminLink tries to store our public key in the conode
+func adminLink(c *cli.Context) error {
+	if pin := c.String("pin"); pin != "" {
+		return errors.New("pin entry not implemented yet")
+	}
+	private := c.String("private")
+	if private == "" {
+		return errors.New("give either pin or private.toml of conode")
+	}
+	var remote struct {
+		Private string
+		Public  string
+		Address network.Address
+	}
+	_, err := toml.DecodeFile(private, &remote)
+	if err != nil {
+		return errors.New("error while reading private.toml: " + err.Error())
+	}
+	pkey, err := encoding.StringHexToScalar(skipchain.Suite, remote.Private)
+	if err != nil {
+		return errors.New("couldn't decode private key: " + err.Error())
+	}
+	pubkey, err := encoding.StringHexToPoint(skipchain.Suite, remote.Public)
+	if err != nil {
+		return errors.New("couldn't decode public key: " + err.Error())
+	}
+	cfg := getConfigOrFail(c)
+	si := network.NewServerIdentity(pubkey, remote.Address)
+	cfg.Link[si.Public.String()] = &link{
+		Private: pkey,
+		Address: remote.Address,
+		Conode:  si,
+	}
+	cerr := skipchain.NewClient().CreateLinkPrivate(si, pkey, skipchain.Suite.Point().Mul(pkey, nil))
+	if cerr != nil {
+		log.Error(cerr)
+		return cerr
+	}
+	log.Lvl1("Correctly linked with", remote.Address)
+	return cfg.save(c)
+}
+
+func adminAuth(c *cli.Context) error {
+	if c.NArg() < 2 {
+		return errors.New("give the (public_hex|IP:Port) and level as argument")
+	}
+	lvl, err := strconv.Atoi(c.Args().Get(1))
+	if err != nil {
+		return err
+	}
+	cfg := getConfigOrFail(c)
+	if len(cfg.Link) == 0 {
+		return errors.New("no link yet")
+	}
+	publicOrIP := c.Args().First()
+	// Check if it's a public key and if we can find it
+	other := cfg.Link[publicOrIP]
+	if other == nil {
+		var err error
+		other, err = findLinkFromAddress(cfg, publicOrIP)
+		if err != nil {
+			return errors.New("couldn't parse address: " + err.Error())
+		}
+	}
+	if other == nil {
+		return errors.New("didn't find host with ip or public key: " + publicOrIP)
+	}
+	return skipchain.NewClient().SettingAuthentication(other.Conode, other.Private, lvl)
```

Comment:
I tried to set level 2 and I got "websocket: close 4103: this authentication type is not yet supported"

Reply:
Implemented differently now.